// @azure/storage-queue/dist-esm/src/generated/src/models/index.js
/*
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is regenerated.
 */
/** Known values of {@link StorageErrorCode} that the service accepts. */
export var KnownStorageErrorCode;
(function (KnownStorageErrorCode) {
    KnownStorageErrorCode["AccountAlreadyExists"] = "AccountAlreadyExists";
    KnownStorageErrorCode["AccountBeingCreated"] = "AccountBeingCreated";
    KnownStorageErrorCode["AccountIsDisabled"] = "AccountIsDisabled";
    KnownStorageErrorCode["AuthenticationFailed"] = "AuthenticationFailed";
    KnownStorageErrorCode["AuthorizationFailure"] = "AuthorizationFailure";
    KnownStorageErrorCode["ConditionHeadersNotSupported"] = "ConditionHeadersNotSupported";
    KnownStorageErrorCode["ConditionNotMet"] = "ConditionNotMet";
    KnownStorageErrorCode["EmptyMetadataKey"] = "EmptyMetadataKey";
    KnownStorageErrorCode["InsufficientAccountPermissions"] = "InsufficientAccountPermissions";
    KnownStorageErrorCode["InternalError"] = "InternalError";
    KnownStorageErrorCode["InvalidAuthenticationInfo"] = "InvalidAuthenticationInfo";
    KnownStorageErrorCode["InvalidHeaderValue"] = "InvalidHeaderValue";
    KnownStorageErrorCode["InvalidHttpVerb"] = "InvalidHttpVerb";
    KnownStorageErrorCode["InvalidInput"] = "InvalidInput";
    KnownStorageErrorCode["InvalidMd5"] = "InvalidMd5";
    KnownStorageErrorCode["InvalidMetadata"] = "InvalidMetadata";
    KnownStorageErrorCode["InvalidQueryParameterValue"] = "InvalidQueryParameterValue";
    KnownStorageErrorCode["InvalidRange"] = "InvalidRange";
    KnownStorageErrorCode["InvalidResourceName"] = "InvalidResourceName";
    KnownStorageErrorCode["InvalidUri"] = "InvalidUri";
    KnownStorageErrorCode["InvalidXmlDocument"] = "InvalidXmlDocument";
    KnownStorageErrorCode["InvalidXmlNodeValue"] = "InvalidXmlNodeValue";
    KnownStorageErrorCode["Md5Mismatch"] = "Md5Mismatch";
    KnownStorageErrorCode["MetadataTooLarge"] = "MetadataTooLarge";
    KnownStorageErrorCode["MissingContentLengthHeader"] = "MissingContentLengthHeader";
    KnownStorageErrorCode["MissingRequiredQueryParameter"] = "MissingRequiredQueryParameter";
    KnownStorageErrorCode["MissingRequiredHeader"] = "MissingRequiredHeader";
    KnownStorageErrorCode["MissingRequiredXmlNode"] = "MissingRequiredXmlNode";
    KnownStorageErrorCode["MultipleConditionHeadersNotSupported"] = "MultipleConditionHeadersNotSupported";
    KnownStorageErrorCode["OperationTimedOut"] = "OperationTimedOut";
    KnownStorageErrorCode["OutOfRangeInput"] = "OutOfRangeInput";
    KnownStorageErrorCode["OutOfRangeQueryParameterValue"] = "OutOfRangeQueryParameterValue";
    KnownStorageErrorCode["RequestBodyTooLarge"] = "RequestBodyTooLarge";
    KnownStorageErrorCode["ResourceTypeMismatch"] = "ResourceTypeMismatch";
    KnownStorageErrorCode["RequestUrlFailedToParse"] = "RequestUrlFailedToParse";
    KnownStorageErrorCode["ResourceAlreadyExists"] = "ResourceAlreadyExists";
    KnownStorageErrorCode["ResourceNotFound"] = "ResourceNotFound";
    KnownStorageErrorCode["ServerBusy"] = "ServerBusy";
    KnownStorageErrorCode["UnsupportedHeader"] = "UnsupportedHeader";
    KnownStorageErrorCode["UnsupportedXmlNode"] = "UnsupportedXmlNode";
    KnownStorageErrorCode["UnsupportedQueryParameter"] = "UnsupportedQueryParameter";
    KnownStorageErrorCode["UnsupportedHttpVerb"] = "UnsupportedHttpVerb";
    KnownStorageErrorCode["InvalidMarker"] = "InvalidMarker";
    KnownStorageErrorCode["MessageNotFound"] = "MessageNotFound";
    KnownStorageErrorCode["MessageTooLarge"] = "MessageTooLarge";
    KnownStorageErrorCode["PopReceiptMismatch"] = "PopReceiptMismatch";
    KnownStorageErrorCode["QueueAlreadyExists"] = "QueueAlreadyExists";
    KnownStorageErrorCode["QueueBeingDeleted"] = "QueueBeingDeleted";
    KnownStorageErrorCode["QueueDisabled"] = "QueueDisabled";
    KnownStorageErrorCode["QueueNotEmpty"] = "QueueNotEmpty";
    KnownStorageErrorCode["QueueNotFound"] = "QueueNotFound";
    KnownStorageErrorCode["AuthorizationSourceIPMismatch"] = "AuthorizationSourceIPMismatch";
    KnownStorageErrorCode["AuthorizationProtocolMismatch"] = "AuthorizationProtocolMismatch";
    KnownStorageErrorCode["AuthorizationPermissionMismatch"] = "AuthorizationPermissionMismatch";
    KnownStorageErrorCode["AuthorizationServiceMismatch"] = "AuthorizationServiceMismatch";
    KnownStorageErrorCode["AuthorizationResourceTypeMismatch"] = "AuthorizationResourceTypeMismatch";
    KnownStorageErrorCode["FeatureVersionMismatch"] = "FeatureVersionMismatch";
})(KnownStorageErrorCode || (KnownStorageErrorCode = {}));
